iotdb-core/datanode/src/main/java/org/apache/iotdb/db/storageengine/dataregion/DataRegion.java

```diff
@@ -2005,39 +1986,47 @@ private void deleteDataInFiles(
       if (tsFileResource.isClosed()) {
-        // delete data in sealed file
-        if (tsFileResource.isCompacting()) {
-          // we have to set modification offset to MAX_VALUE, as the offset of source chunk may
-          // change after compaction
-          deletion.setFileOffset(Long.MAX_VALUE);
-          // write deletion into compaction modification file
-          tsFileResource.getCompactionModFile().write(deletion);
-          // write deletion into modification file to enable read during compaction
-          modFile.write(deletion);
-          // remember to close mod file
-          tsFileResource.getCompactionModFile().close();
-          modFile.close();
-        } else {
-          deletion.setFileOffset(tsFileResource.getTsFileSize());
-          // write deletion into modification file
-          boolean modFileExists = modFile.exists();
-          long originSize = modFile.getSize();
-          modFile.write(deletion);
+        try {
+          tsFileResource.writeLock();
+          // delete data in sealed file
+          if (tsFileResource.isCompacting()) {
+            // we have to set modification offset to MAX_VALUE, as the offset of source chunk may
+            // change after compaction
+            deletion.setFileOffset(Long.MAX_VALUE);
+            // write deletion into compaction modification file
+            tsFileResource.getCompactionModFile().write(deletion);
+            // write deletion into modification file to enable read during compaction
+            modFile.write(deletion);
+            // remember to close mod file
+            tsFileResource.getCompactionModFile().close();
+            modFile.close();
+          } else {
+            deletion.setFileOffset(tsFileResource.getTsFileSize());
+            // write deletion into modification file
+            boolean modFileExists = modFile.exists();
 
-          // remember to close mod file
-          modFile.close();
+            modFile.write(deletion);
 
-          // if file length greater than 1M,execute compact.
-          modFile.compact();
+            // remember to close mod file
+            modFile.close();
 
-          if (!modFileExists) {
-            FileMetrics.getInstance().increaseModFileNum(1);
-          }
+            // if file length greater than 1M,execute compact.
+            modFile.compact();
+
+            if (!modFileExists) {
+              FileMetrics.getInstance().increaseModFileNum(1);
+            }
 
-          // The file size may be smaller than the original file, so the increment here may be
-          // negative
-          FileMetrics.getInstance().increaseModFileSize(modFile.getSize() - originSize);
+            // The file size may be smaller than the original file, so the increment here may be
+            // negative
+            FileMetrics.getInstance().increaseModFileSize(modFile.getSize() - originSize);
+          }
+        } catch (IOException e) {
+          modFile.truncate(originSize);
```

Comment:
```suggestion
        } catch (Throwable t) {
          if (originSize != -1) {
            modFile.truncate(originSize);
          }
          throw t;
```

We must rethrow the exception, otherwise this excption will be ignored, we should at least let user know that their deletion operation failed.